x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,391 @@
+	return external.LoadDefaultAWSConfig(
+		external.WithSharedConfigProfile(config.ProfileName),
+	)
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.logger.Debugf("s3", "Run s3 input with queueURLs: %+v", p.config.QueueURLs)
+	visibilityTimeout := int64(p.config.VisibilityTimeout)
+	for _, queueURL := range p.config.QueueURLs {
+		regionName, err := getRegionFromQueueURL(queueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %s", queueURL)
+			continue
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// update message visibility timeout if it's taking longer than 1/2 of
+		// visibilityTimeout to make sure filebeat can finish reading
+		changeMessageVisibilityInput := &sqs.ChangeMessageVisibilityInput{
+			QueueUrl:          &queueURL,
+			VisibilityTimeout: &visibilityTimeout,
+		}
+
+		// receive messages
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &queueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+		output, err := req.Send()
+		if err != nil {
+			p.logger.Error("failed to receive message from SQS:", err)
+			continue
+		}
+
+		// process messages received from sqs
+		if len(output.Messages) > 0 {
+			var wg sync.WaitGroup
+			numMessages := len(output.Messages)
+			wg.Add(numMessages)
+
+			for i := range output.Messages {
+				done := make(chan struct{})
+				// launch goroutine to handle each message from sqs
+				go func(message sqs.Message) {
+					defer wg.Done()
+					defer close(done)
+
+					s3Infos, err := handleMessage(message)
+					if err != nil {
+						p.logger.Error(err.Error())
+					}
+
+					// read from s3 object and create event for each log line
+					p.readS3Object(svcS3, s3Infos)
+
+					// delete message after events are sent
+					err = deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+					if err != nil {
+						p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+					}
+				}(output.Messages[i])
+
+				go func(message sqs.Message) {
+					for {
+						select {
+						case <-done:
+							return
+						case <-time.After(time.Duration(visibilityTimeout/2) * time.Second):
+							// if half of the set visibilityTimeout passed and this is
+							// still ongoing, then change visibility timeout.
+							changeMessageVisibilityInput.ReceiptHandle = message.ReceiptHandle
+							req := svcSQS.ChangeMessageVisibilityRequest(changeMessageVisibilityInput)
+							_, err = req.Send()
+							if err != nil {
+								p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+							}
+						}
+					}
+				}(output.Messages[i])
+			}
+			wg.Wait()
+		}
+	}
+}
+
+// Stop stops the input and all its harvesters
+func (p *Input) Stop() {
+	p.outlet.Close()
```

Comment:
Sorry I only saw this now. Calling stop here won't stop the read loop in `Run`? this would need a close mechanism

Reply:
Yes you are right! Please let me know if the new `Stop` function looks good to you.